cmd/clusterctl/client/tree/discovery.go

```diff
@@ -83,19 +84,22 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 	tree := NewObjectTree(cluster, options.toObjectTreeOptions())
 
 	// Adds cluster infra
-	if clusterInfra, err := external.Get(ctx, c, cluster.Spec.InfrastructureRef, cluster.Namespace); err == nil {
-		tree.Add(cluster, clusterInfra, ObjectMetaName("ClusterInfrastructure"))
+	clusterInfra, err := external.Get(ctx, c, cluster.Spec.InfrastructureRef, cluster.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("get InfraCluster reference from Cluster: %v", err)
 	}
+	tree.Add(cluster, clusterInfra, ObjectMetaName("ClusterInfrastructure"))
 
 	if options.ShowClusterResourceSets {
 		addClusterResourceSetsToObjectTree(ctx, c, cluster, tree)
 	}
 
 	// Adds control plane
 	controlPlane, err := external.Get(ctx, c, cluster.Spec.ControlPlaneRef, cluster.Namespace)
-	if err == nil {
-		addControlPlane(cluster, controlPlane, tree, options)
+	if err != nil {
```

Comment:
This should not be changed because CAPI still supports clusters with standalone control plane machines (and thus without a control plane object)